lib/toolbox/stream/stream.c

```diff
@@ -0,0 +1,337 @@
+    return stream_delete_and_insert(stream, delete_size, stream_write_struct, &write_data);
+}
+
+bool stream_delete_and_insert_cstring(Stream* stream, size_t delete_size, const char* string) {
+    furi_assert(stream);
+    StreamWriteData write_data = {.data = (uint8_t*)string, .size = strlen(string)};
+    return stream_delete_and_insert(stream, delete_size, stream_write_struct, &write_data);
+}
+
+bool stream_delete_and_insert_format(Stream* stream, size_t delete_size, const char* format, ...) {
+    furi_assert(stream);
+    va_list args;
+    va_start(args, format);
+    bool result = stream_delete_and_insert_vaformat(stream, delete_size, format, args);
+    va_end(args);
+
+    return result;
+}
+
+bool stream_delete_and_insert_vaformat(
+    Stream* stream,
+    size_t delete_size,
+    const char* format,
+    va_list args) {
+    furi_assert(stream);
+    string_t data;
+    string_init_vprintf(data, format, args);
+    StreamWriteData write_data = {
+        .data = (uint8_t*)string_get_cstr(data), .size = string_size(data)};
+    bool result = stream_delete_and_insert(stream, 0, stream_write_struct, &write_data);
+    string_clear(data);
+
+    return result;
+}
+
+bool stream_delete(Stream* stream, size_t size) {
+    furi_assert(stream);
+    return stream_delete_and_insert(stream, size, NULL, NULL);
+}
+
+size_t stream_copy(Stream* stream_from, Stream* stream_to, size_t size) {
+    uint8_t* buffer = malloc(STREAM_CACHE_SIZE);
+    size_t copied = 0;
+
+    do {
+        size_t bytes_count = MIN(STREAM_CACHE_SIZE, size - copied);
+        if(bytes_count <= 0) {
+            break;
+        }
+
+        uint16_t bytes_were_read = stream_read(stream_from, buffer, bytes_count);
+        if(bytes_were_read != bytes_count) break;
+
+        uint16_t bytes_were_written = stream_write(stream_to, buffer, bytes_count);
+        if(bytes_were_written != bytes_count) break;
+
+        copied += bytes_count;
+    } while(true);
+
+    free(buffer);
+    return copied;
+}
+
+size_t stream_copy_full(Stream* stream_from, Stream* stream_to) {
+    size_t was_written = 0;
+
+    do {
+        if(!stream_seek(stream_from, 0, StreamOffsetFromStart)) break;
+        if(!stream_seek(stream_to, 0, StreamOffsetFromStart)) break;
+        was_written = stream_copy(stream_from, stream_to, stream_size(stream_from));
+    } while(false);
+
+    return was_written;
+}
+
+bool stream_split(Stream* stream, Stream* stream_left, Stream* stream_right) {
+    bool result = false;
+    size_t size = stream_size(stream);
+    size_t tell = stream_tell(stream);
+
+    do {
+        // copy right
+        if(stream_copy(stream, stream_right, size - tell) != (size - tell)) break;
+
+        // copy left
+        if(!stream_rewind(stream)) break;
+        if(stream_copy(stream, stream_left, tell) != tell) break;
+
+        // restore RW pointer
+        if(!stream_seek(stream, tell, StreamOffsetFromStart)) break;
+        result = true;
+    } while(false);
```

Comment:
no cleanup - so can just `return true`, without `do..while` (imo)

Reply:
"do {} while (0)" makes it easier to maintain the code in the future